src/browser.c

```diff
@@ -0,0 +1,147 @@
+#include "browser.h"
+
+#include <stdio.h>
+#include <string.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fts.h>
+#include <errno.h>
+#include <limits.h>
+#include <unistd.h>
+
+struct browser {
+    /** File list. */
+    char** files;
+    int max;
+    int total;
+    int current;
+    bool recursive;   ///< Recurse into subdirectories
+};
+
+void add_file(browser* browser, const char* file)
+{
+    if (browser->max == browser->total) {
+        browser->max *= 2;
+        browser->files = realloc(browser->files, browser->max * sizeof(char*));
+        if (!browser->files) {
+            fprintf(stderr, "Not enough memory\n");
+            return;
+        }
+    }
+    size_t len = strlen(file);
+    browser->files[browser->total] = malloc(len + 1);
+    memcpy(browser->files[browser->total], file, len);
+    browser->files[browser->total][len] = '\0';
```

Comment:
I don't like `strcpy`/`strncpy`, because they internally call `strlen` or similar function, which is useless, since we already know the length of the string. If you prefer readability before speed, I can of course change it.